### PR TITLE
Update the download_one task to use the streamer.

### PR DESCRIPTION
**Pull request by jeremycline**
This also updates the streamer to check for a special HTTP header to
determine whether or not it should dispatch a task and changes the
streamer to use the new LazyCatalogEntry model.

I'm hesitant to write tests for the task just yet, since we may end up changing it quite a bit depending on whether we download entire content units or not.

### CODE REVIEW

File: server/pulp/server/controllers/content.py [hunk 1]
```diff
@@ -111,83 +126,110 @@ def refresh_content_source(content_source_id=None):
...
+    scheme = 'https'
+    host = pulp_conf.get('lazy', 'redirect_host')
+    port = pulp_conf.get('lazy', 'redirect_port')
+    path_prefix = pulp_conf.get('lazy', 'redirect_path')
+    return ContentView.urljoin(scheme, host, port, path_prefix,
+                               catalog_entry.path, '')
 
 
-def queue_download(plugin_id, unit_locator, url, options):
+def queue_download(catalog_entry):
```
Comment (jortel) [inline server/pulp/server/controllers/content.py]: should this be `queue_download_one()`?

File: server/pulp/server/controllers/content.py [hunk 2]
```diff
@@ -111,83 +126,110 @@ def refresh_content_source(content_source_id=None):
...
+    :param catalog_entry: The catalog entry to get the URL for.
+    :type  catalog_entry: pulp.server.db.model.LazyCatalogEntry
+
+    :return: The streamer URL which corresponds to the content unit.
+    """
+    scheme = 'https'
+    host = pulp_conf.get('lazy', 'redirect_host')
+    port = pulp_conf.get('lazy', 'redirect_port')
+    path_prefix = pulp_conf.get('lazy', 'redirect_path')
+    return ContentView.urljoin(scheme, host, port, path_prefix,
```
Comment (jortel) [inline server/pulp/server/controllers/content.py]: Shouldn't this be be a signed URL?

File: server/pulp/server/constants.py [hunk 3]
```diff
@@ -2,3 +2,5 @@
 
 PULP_USER_METADATA_FIELDNAME = 'pulp_user_metadata'
 PULP_DJANGO_SETTINGS_MODULE = 'pulp.server.webservices.settings'
+
+PULP_STREAM_REQUEST_HEADER = 'Pulp-Stream-Request'
```
Comment (jortel) [inline server/pulp/server/constants.py]: I wonder if this should begin with 'X-'?

File: server/pulp/server/controllers/content.py [hunk 4]
```diff
@@ -111,83 +126,110 @@ def refresh_content_source(content_source_id=None):
...
-    :type url: str
-    :param options: Passed to the Importer.get_downloader().
-    :type options: dict
+    :param catalog_entry: The catalog entry to download content for.
+    :type  catalog_entry: pulp.server.db.model.LazyCatalogEntry
     """
-    _logger.info(_('Downloading: {url}').format(url=url))
+    _logger.info(_('Downloading {url} via the Pulp streamer').format(url=catalog_entry.url))
 
     # Find the plugin
```
Comment (jortel) [inline server/pulp/server/controllers/content.py]: Comment seems unrelated.

File: server/pulp/server/lazy/streamer.py [hunk 5]
```diff
@@ -102,51 +103,64 @@ def render_GET(self, request):
...
         :param request: The content request.
         :type  request: twisted.web.server.Request
         """
-        relative_path = urlparse(request.uri).path.lstrip('/')
+        catalog_path = urlparse(request.uri).path
         with Responder(request) as responder:
             try:
                 catalog_entry = model.LazyCatalogEntry.objects(
-                    relative_path=relative_path).order_by('-plugin_id').first()
+                    path=catalog_path).order_by('-plugin_id').first()
```
Comment (jortel) [inline server/pulp/server/lazy/streamer.py]: s/-plugin_id/importer_id/

File: server/pulp/server/lazy/streamer.py [hunk 6]
```diff
@@ -102,51 +103,64 @@ def render_GET(self, request):
...
-                downloader.event_listener = StreamerListener(request)
-                downloader.download_one(download_request, events=True)
-
-                # Dispatch the Pulp task to download the unit.
-                content_controller.queue_download(catalog_entry.plugin_id,
-                                                  catalog_entry.unit_locator,
-                                                  catalog_entry.url,
-                                                  catalog_entry.data)
+                # Only dispatch the task if the request didn't originate from Pulp.
+                if not request.getHeader(PULP_STREAM_REQUEST_HEADER):
```
Comment (jortel) [inline server/pulp/server/lazy/streamer.py]: :+1:

File: server/pulp/server/controllers/content.py (hunk 2 above)
Comment (jeremycline) [inline server/pulp/server/controllers/content.py]: Oops,  it definitely should be.

File: server/pulp/server/constants.py (hunk 3 above)
Comment (jeremycline) [inline server/pulp/server/constants.py]: I did a bit of research about this because I knew at one point non-standard headers were supposed to start with 'X-'. According to http://tools.ietf.org/html/rfc6648#section-3 new headers should not use the 'X-'.
